### ISSUE TITLE
zero padding in mrcnn_class_loss_graph

### ISSUE THREAD

**Issue by KLH1472** [matterport/Mask_RCNN#972]
```
def mrcnn_class_loss_graph(target_class_ids, pred_class_logits,
                           active_class_ids):
    """Loss for the classifier head of Mask RCNN.

    target_class_ids: [batch, num_rois]. Integer class IDs. Uses zero
        padding to fill in the array.
    pred_class_logits: [batch, num_rois, num_classes] # shape=(?, 200, 81)
    active_class_ids: [batch, num_classes]. Has a value of 1 for
        classes that are in the dataset of the image, and 0
        for classes that are not in the dataset.
    """
    # During model building, Keras calls this function with
    # target_class_ids of type float32. Unclear why. Cast it
    # to int to get around it.
    target_class_ids = tf.cast(target_class_ids, 'int64')

    # Find predictions of classes that are not in the dataset.
    pred_class_ids = tf.argmax(pred_class_logits, axis=2)
    # TODO: Update this line to work with batch > 1. Right now it assumes all
    #       images in a batch have the same active_class_ids
    pred_active = tf.gather(active_class_ids[0], pred_class_ids)

    # Loss
    loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
        labels=target_class_ids, logits=pred_class_logits)

    # Erase losses of predictions of classes that are not in the active
    # classes of the image.
    loss = loss * pred_active

    # Computer loss mean. Use only predictions that contribute
    # to the loss to get a correct mean.
    loss = tf.reduce_sum(loss) / tf.reduce_sum(pred_active)
    return loss
```
I noted that positive ROIs, negative ROIs and zero padding ROIs all contribute to this loss. Should zero padding contributes to the loss? Is it better that we only use positive ROIs, negative ROIs to calculate the loss? Thanks for your suggestion!

**Comment by KLH1472** [comment]
In inference, fpn_classifier_graph will have a chance to meet a patch cropped from an image by 'zero padding ROIs'. So it's necessary to train fpn_classifier_graph to classify this kind of patches as 'background'.
Is it right?

**Comment by qjfytz** [comment]
I have the same question.  In my opinion, since the classification losses of padding rois are always zeros, it actually add a weight parameter (approximate number_of_roi/(number_of_roi + number_of_padding), which is less than 1) on the mrcnn_class_loss.  

So is this operation reasonable?

**Comment by bertsky** [comment]
@KLH1472 your observation is correct. Patch/crop mode or images with relatively small regions or empty images are all problematic in this implementation, because if there are no positive proposals then there will be no negatives as well and the network needs to learn turning padded zeros into BG. This is more fragile than learning true zeros for BG because if during inference the RPN yields enough proposals to fill a batch then there will be many false positive detection. See [here](https://github.com/matterport/Mask_RCNN/pull/1088#issuecomment-695208655) for a solution.

**Comment by rcx986635** [comment]
@waleedka @KLH1472 @bertsky  Finaly, Is there an exact answer to the question  "Should zero padding contributes to the loss? Is it better if we only use positive ROIs, negative ROIs to calculate the loss?", Thank you for your replies.
